### ISSUE TITLE
how to handle the first simulation step

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#1195]
**Reported by lennart&#x2e;ochel on 2 Jul 2013 10:09 UTC**
I am not sure if the specification says what the correct input for the first simulation step is. Either I cannot find it or it is just not there. The issue can be easily shown with the following model:
```mo
model Unnamed
  Real x(start=0, fixed=true);
  Real d(start=0, fixed=true);
initial equation
  pre(x) = -1;
equation
  der(x) = 0;
  when sample(0.0, 1.0) then
    d = pre(x);
  end when;
end Unnamed;
```
I guess, it is quite simple to solve the initialization problem for this. It should be the following (this is also what OpenModelica does):
```
LOG_SOTI          | info    | ### SOLUTION OF THE INITIALIZATION ###
|                 | |       | | states variables
|                 | |       | | | [1] Real x(start=0, nominal=1) = 0 (pre: -1)
|                 | |       | | derivatives variables
|                 | |       | | | [2] Real der(x) = 0
|                 | |       | | other real variables
|                 | |       | | | [3] Real d(start=0, nominal=1) = 0 (pre: 0)
```
What I do still not understand is how the first simulation step is treated. Do I need to override each pre value with its actual value like it should be done after event iteration? Therefore `d` should be `0` for the simulation. But I guess this is wrong due to the specification, because the initialization is not an event iteration. Anyway, this is what we do currently in OpenModelica.
If the pre-values gets not overridden, the sample event at `t=0` forces d to be `1`. I guess, this is the correct behavior due to the current specification. Anyway, the specification is very imprecise in that point.

I can only test this with Dymola 2014 and OpenModelica:
OpenModelica: initialization works as expected (see above) and `d` is `0` for the first time unit.
Dymola 2014: initialization fails (the error message suggests that Dymola handles the initialization problem weird and probably wrong):
```
The initialization problem is overspecified for variables of element type Real.

The initial equation 
x_.start = -1;
refers to variables, which all are known.
To correct it you can remove this equation or inactivate start values for 
x_.start(start = 0)
```

Can anyone point out where this is handled in the specification?
(this ticket continues and focus the discussion that starts here: #1169)

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1195

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 2 Jul 2013 12:33 UTC**
I discussed this issue last week with Sven Erik and yesterday with Hans. It seems we all have the same opinion on this topic. Let me summarize this as follow:

The current Modelica Specification on this topic is incomplete and should be fixed. Here is our proposal:

**Observations**

Informally, a variable v in Modelica can have the following "values":

 * v: The actual value of a variable (= "right limit").
 * v.start: The start value of a variable used for initialization (see below).
 * pre(v): The value of the variable from the previous model evaluation (= "left limit").

Additionally, there are the following properties defined in the specification:

 * For a discrete variable, the following constraint holds: v.start = pre(v)
 * For a non-discrete variable, the following constraint holds: v.start = v

This means that for a discrete variable, v.start and pre(v) is the same during initialization,
so this is well-defined.

For non-discrete variables, the relationship between v.start and pre(v) seems to be not defined in the
specification.

Intuitively, they should be the same also for non-discrete variables, so v.start = pre(v) during initialization:

 * If a non-discrete variable is without any discontinuity, then v = pre(v) holds all the time and also at initial time (therefore also: v.start = pre(v)).
  
 * If there is a discontinuity between the value of initialization and the value at the first event, which happens to be the initial time, then clearly at this event, pre(v) is the value of the variable during initialization, so v.start.
   
 * The only use-case where pre(v) and v.start might be different during initialization is for reinit, such as, "when initial() then reinit(v, -e\*pre(v)) end when;" and therefore, it might make sense to define different values for v.start and pre(v). However, this could also be formulated (with the new reinit definition) as: "when initial() then reinit(v, -e\*v)    end when;" (meaning that "new(v) = -e\*v.start") which is unambiguous.

To summarize, there seems to be now use case where different values for v.start and pre(v)
makes sense during initialization.

**Proposal**
Based on the observations above, the following additional rule is introduced in to the specification:

Treat the start attribute of a variable v (conceptually) as an independent variable v.start
during initialization. During initialization, the following equations are implicitely present:

 * For all variables v: pre(v) = v.start.
 * For all non-discrete variables r, additionally: r = r.start (this means: r = r.start = pre(r))

Furthermore, the update equation pre(v) := v is only evaluated after every model evaluation at
an event instant, but not during initialization (otherwise, v = pre(v) always after initialization,
but in some cases this is not desired; if this is desired, one can formulate it explicitly with the
initialization equations).

**Example**
With these rules, the example is not correct because we have 4 equations:

```
initial equation
   x.start = 0
   pre(x) = -1
   pre(x) = x.start   // implicitly present
   x = x.start        // implicitly present
```

for three unknowns (x, x.start, pre(x)) during initialization.

If one of the initial conditions for x is removed, say x.start, everything is well defined for x.
In this case after initializiation (and before the first event at time = 0), x = pre(x) = -1.

"d" is also well-defined since

 * the when-clause is not active during initialization and therefore during initialization: d = pre(d) = d.start = 0
 * at the first event at time = 0 we have d = pre(x) = -1

**Comment by modelica-trac-importer** [comment]
**Comment by lennart&#46;ochel on 2 Jul 2013 14:59 UTC**
Replying to [comment:1 otter]:
> Additionally, there are the following properties defined in the specification:
> 
>  \* For a discrete variable, the following constraint holds: v.start = pre(v)
>  \* For a non-discrete variable, the following constraint holds: v.start = v
> 
> This means that for a discrete variable, v.start and pre(v) is the same during initialization,
> so this is well-defined.

I guess this is a wrong observation. This constraints exist only with additionally fixed=true attribute. But probably I am wrong ...
Where is it in the specification?

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 2 Jul 2013 15:21 UTC**
Replying to [comment:2 lennart.ochel@…]:


> Replying to [comment:1 otter]:
> 
> > Additionally, there are the following properties defined in the specification:
> > 
> > \* For a discrete variable, the following constraint holds: v.start = pre(v)
> > \* For a non-discrete variable, the following constraint holds: v.start = v
> > 
> > This means that for a discrete variable, v.start and pre(v) is the same during initialization,
> > so this is well-defined.
> > 
> 
> I guess this is a wrong observation. This constraints exist only with additionally fixed=true attribute. But probably I am wrong ...
> Where is it in the specification?

Sorry, this is what I meant: Whenever v.start is mentioned, this only holds if fixed=true.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 3 Jul 2013 12:53 UTC**
Replying to [comment:1 otter]:

> I discussed this issue last week with Sven Erik and yesterday with Hans. It seems we all have the same opinion on this topic. Let me summarize this as follow:
> ....
> 
> 
> **Proposal**
> Based on the observations above, the following additional rule is introduced in to the specification:
> 
> Treat the start attribute of a variable v (conceptually) as an independent variable v.start
> during initialization. During initialization, the following equations are implicitely present:
> 
> \* For all variables v: pre(v) = v.start.
> \* For all non-discrete variables r, additionally: r = r.start (this means: r = r.start = pre(r))
> 
> Furthermore, the update equation pre(v) := v is only evaluated after every model evaluation at
> an event instant, but not during initialization (otherwise, v = pre(v) always after initialization,
> but in some cases this is not desired; if this is desired, one can formulate it explicitly with the
> initialization equations).
> 

Here is the proposal in a more precise way:

If attribute fixed=true is defined for a variable v and pre(v) appears in the class where v is declared or in a class that inherits from this class, then conceptually the equation

```
initial equation
  pre(v) = v.start
```
is present.

If attribute fixed=true is defined for the non-discrete variable r, then conceptually the equation

```
initial equation
  r = r.start
```
is present.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 3 Jul 2013 12:57 UTC**
Changed milestone to ModelicaSpec3.2rev2, since this need to be clarified before MSL 3.2.1 can be released (model Modelica.Blocks.Math.Mean).

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 3 Jul 2013 13:15 UTC**
Replying to [comment:4 otter]:
> Here is the proposal in a more precise way:
> 
> If attribute fixed=true is defined for a variable v and pre(v) appears in the class where v is declared or in a class that inherits from this class, then conceptually the equation

If the generation of an equation depends from the source of pre(v), then this is not well defined. What about other operators, having the pre implicit (edge, change, when). What is the reason for that?
> 
> ```
> initial equation
>   pre(v) = v.start
> ```
> is present.
> 

This equation should be generated if v is discrete and fixed=true, period.

I found your former description very clear, but this comment creates some confusion.

**Comment by modelica-trac-importer** [comment]
**Comment by lennart&#x2e;ochel on 3 Jul 2013 13:35 UTC**
I guess there is no need to define additional initial equations that are implicit present if fixed=true.
The existing specification is quite clear how to solve the initialization problem. The need of providing also pre-values for some continuous variables depends on the model and should be done by the modeler. In most cases this is not necessary. Anyway, this has to be defined clearly without a connection to the fixed attribute.

The following model should behave exactly the same than the one above:
```mo
model Unnamed
  Real x(start=0);
  Real d(start=0);
initial equation 
  pre(x) = -1;
  x = 0;
  pre(d) = 0;
equation 
  der(x) = 0;
  when sample(0.0, 1.0) then
    d = pre(x);
  end when;
end Unnamed;
```

**Comment by modelica-trac-importer** [comment]
**Comment by lennart&#46;ochel on 3 Jul 2013 13:44 UTC**
Replying to [comment:1 otter]:
> 
> "d" is also well-defined since
> 
>  \* the when-clause is not active during initialization and therefore during initialization: d = pre(d) = d.start = 0
>  \* at the first event at time = 0 we have d = pre(x) = -1

I agree. `d` should be `-1` at the first event at time = 0.

If I simulate this model with Dymola, `d` is `0` for the first time unit. In my opinion this is wrong.

What there should be clarified is only that the solution of the initialization is used to evaluate the first simulation step (including pre values and actual values). With this information everything is well-definied.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 3 Jul 2013 14:13 UTC**
Replying to [comment:8 lennart.ochel@…]:
> I guess there is no need to define additional initial equations that are implicit present if fixed=true.
> The existing specification is quite clear how to solve the initialization problem. The need of providing also pre-values for some continuous variables depends on the model and should be done by the modeler. In most cases this is not necessary. Anyway, this has to be defined clearly without a connection to the fixed attribute.
> 
> The following model should behave exactly the same than the one above:
> ```mo
> model Unnamed
>   Real x(start=0);
>   Real d(start=0);
> initial equation 
>   pre(x) = -1;
>   x = 0;
>   pre(d) = 0;
> equation 
>   der(x) = 0;
>   when sample(0.0, 1.0) then
>     d = pre(x);
>   end when;
> end Unnamed;
> ```

Sorry, this does not make sense and is not practical, and indicate that we need a different formulation - **when fixed=false**.

Basically 'x' is a continuous-time state and we need to provide exactly one initial condition for it. Saying that we need either one start-value (with fixed=true) or two initial equations will not work well for users.

Next consider
```
model Unnamed
  Real x=foo(time);
  Real d(start=0);
initial equation 
  pre(x) = -1;
  pre(d) = 0;
equation 
  when sample(0.0, 1.0) then
    d = pre(x);
  end when;
end Unnamed;
```
Basically the sampling of x forces us to introduce an initial equation for the pre-value of x -  and it is not even used!

Thus I propose the following instead (it is the same as Martin Otter's for fixed=true)

If attribute fixed=true is defined for a [discrete or non-discrete] variable v, then conceptually the equation
```
initial equation
  pre(v) = v.start;
```
is present. 

For any non-discrete variable r, then conceptually the equation
```
initial equation
  pre(v) = v;
```
is present. [Typically, if pre(v) is not present in the flattened model we can eliminate this equation. In case the attribute fixed=true we can directly add the initial equation that v=v.start; and in other cases we can just remove pre(v).]

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 4 Jul 2013 05:03 UTC**
If v is discrete, pre(v) is always (atleast implicit) present in a model, because it is needed for the decision whether an event iteration has to be continued. So the omission of the equation applies only in the continuous case. Isn't it?

Another question is, what happens after the initial equation have been evaluated and consistent initial conditions are calculated? Are the pre(v) variables assigned with the value of v and then a usual event iteration follows at start time? Usually, the invariant during transient calculation is pre(v)=v. Only during even steps (and initialization) they can be different. So this has to be assured also after initialization.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 4 Jul 2013 07:49 UTC**
Replying to [comment:11 kurzbach]:
> If v is discrete, pre(v) is always (atleast implicit) present in a model, because it is needed for the decision whether an event iteration has to be continued. So the omission of the equation applies only in the continuous case. Isn't it?

I would believe so.
However, it is an optimization in tools that should have not influence the result - so if someone can find a case where it is safe to perform it for the discrete case as well I see no problem.

> Another question is, what happens after the initial equation have been evaluated and consistent initial conditions are calculated? Are the pre(v) variables assigned with the value of v and then a usual event iteration follows at start time? Usually, the invariant during transient calculation is pre(v)=v. Only during even steps (and initialization) they can be different. So this has to be assured also after initialization.

I agree that e.g. since pre(v) and v can be different during initialization it is usually necessary to have event iteration(s) after the consistent initial conditions are calculated (and additional event iteration should not break anything).

And if pre(v)<>v for the initial conditions one should copy v to pre(v) and perform event iteration (in the usual way until v=pre(v)).

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 4 Jul 2013 09:48 UTC**
That means, when pre(v) is different then v after initialization, there is no way to access it afterwards.
This is a little bit strange, because a calculated value is lost, but also ok for me as long as is somehow (somewhere) described.

**Comment by modelica-trac-importer** [comment]
**Comment by lennart&#46;ochel on 4 Jul 2013 11:03 UTC**
I discussed this today with Bernhard Bachmann. Our thoughts are the following:

**Observations**
During continuous simulation, before any event occurs, it yields pre(v) = v for all continuous and discrete variables. As soon as an event occurs, event iteration is triggered and during this process pre(v) might be different to v. As soon as the event has been handled it yields pre(v) = v again for all variables. _[Why should this be different for an event at time=startTime?]_

Resulting from this, we should get into the first simulation step with pre(v) = v for all variables (continuous and discrete). Anyway, it is necessary to have the possibility to provide different pre values than actual values during initialization (for all variables). This is necessary because it is allowed to enable when clauses for the initialization (using the initial operator) where all pre values can be used. But, this is already possible with the current specification.

_Because the initialization has nothing to do with events, we do not need to perform an event iteration after the initialization._

**Proposal**
After the initial system is solved consistently, all pre values get overridden by their actual values.

**Examples**
The first example is from my initial post:
```mo
model example1
  Real x(start=0, fixed=true);
  Real d(start=0, fixed=true);
initial equation
  pre(x) = -1;
equation
  der(x) = 0;
  when sample(0.0, 1.0) then
    d = pre(x);
  end when;
end example1;
```

Due to the current specification is should behave exactly the same as example1b:
```mo
model example1b
  Real x(start=0);
  Real d(start=0);
initial equation
  pre(x) = -1;
  x = x.start;      // implicitly present in example1
  pre(d) = d.start; // implicitly present in example1
equation
  der(x) = 0;
  when sample(0.0, 1.0) then
    d = pre(x);
  end when;
end example1b;
```

Both examples should simulate with d = 0 due to our proposal. In this case the calculated pre value of x is quite unneeded. Anyway, this is a modeler issue and no fault.

**Remark**
The only change in the specification necessary to end up at a well-defined initialization process and simulation start is the fact that _After the initialization all pre values are overwritten using pre(v):=v_. It seems that all Modelica tools already do this.

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 4 Jul 2013 11:23 UTC**
> Because the initialization has nothing to do with events, we do not need to perform an event iteration after the initialization.

If pre(v) is different to v and you assign pre(v) with v the you change the discrete state of the system, and the equations in the system might not be the same as during initialization. So an event iteration is needed in this case.
So the initialization can be seen as special, very first event step during the event iteration at start time.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 4 Jul 2013 11:42 UTC**
Replying to [comment:15 kurzbach]:
> > Because the initialization has nothing to do with events, we do not need to perform an event iteration after the initialization.
> 
> If pre(v) is different to v and you assign pre(v) with v the you change the discrete state of the system, and the equations in the system might not be the same as during initialization. So an event iteration is needed in this case.
> So the initialization can be seen as special, very first event step during the event iteration at start time.

Was also going to write this; so the event iteration after solving initial conditions is needed in many/most cases, and should always be harmless.

And still my major concern is makes no sense to require either you set one start-value as fixed, or you have to give **two** initial equations! That just adds an unnecessary burden to model developers (and would be a flaw in the language). 

Yes, we were right to change the specification to allow pre(x) in initial equations in 3.2rev2 (resolving #853) - but we should not require it for most variables (both because it would be tedious and because it would break compatibility).

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 4 Jul 2013 11:58 UTC**
Btw. what about calls of ` reinit(...) ` inside ` when initial() `?.
IMO, 
* they should be forbidden, because there are no ODE states during initialization or 
* they have no effect during initialization and are applied (according to our new definition) after that, but before assigning the pre values, which would be consistent with the transient calculations.

The taken choice needs to be in the specification.

**Comment by modelica-trac-importer** [comment]
**Comment by bachmann on 4 Jul 2013 12:23 UTC**
Replying to [comment:15 kurzbach]:
> > Because the initialization has nothing to do with events, we do not need to perform an event iteration after the initialization.
> 
> If pre(v) is different to v and you assign pre(v) with v the you change the discrete state of the system, and the equations in the system might not be the same as during initialization. So an event iteration is needed in this case.
> So the initialization can be seen as special, very first event step during the event iteration at start time.


**This is correct! **

After the initialization has been performed (based on additionally defined initial conditions!) pre(v) might be different to v. But, before the simulation can start pre(v) must be set equal to v, which might force an event iteration. This phase is somehow between the initialization and simulation process!

**Comment by modelica-trac-importer** [comment]
**Comment by lennart&#x2e;ochel on 4 Jul 2013 13:18 UTC**
Yes, `reinit` should have no effect during initialization. This is my current interpretation since reinit takes only effect at an event instant. Also there are no states during initialization.
This should probably also clarified.

Replying to [comment:17 kurzbach]:
> [...] (according to our new definition) [...]

What do you mean by “new definition”? Also Martin Otter mentioned this new definition in his first post.

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 4 Jul 2013 13:39 UTC**
Replying to [comment:19 lennart.ochel@…]:
> Replying to [comment:17 kurzbach]:
> > [...] (according to our new definition) [...]
> What do you mean by “new definition”? Also Martin Otter mentioned this new definition in his first post.

The discussion from the last design meeting is logged in #578. Proposal 3 is the aggreed solution and is put in the specification.

**Comment by modelica-trac-importer** [comment]
**Comment by lennart&#46;ochel on 4 Jul 2013 17:08 UTC**
Here is another try:

**Why should a continuous variable get initialized different to their pre value?**
_[Is there a good reason to do that?]_

I guess this should be avoided. In my opinion there are different reasons for that:
- A continuous variable is continuous!
- reinit is (or should) also be forbidden/ineffective during initialization
- This is the most natural approach (at least for me in the moment ;-))

As a result it should be implicit requested that pre(v) is equal to v for all continuous variables.

After the initial equation system is solved, pre(d) and d might be different in the case of d is discrete. To provide a well-defined starting point for the simulation an event iteration is required to find a consistent system with pre(v) =v for all variables.

If it is necessary for an application to provide a discontinuity for continuous variables (sounds already strange, or not?) at the beginning of the simulation this can be modeled using `when not initial() then reinit(…); end when;`.

As a result my initial model get invalid.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 10 Jul 2013 09:01 UTC**
Trying to analyze the issue:

If we add that during initialization pre(y)=y for non-discrete variables, one possibility is to forbid using pre(y) for non-discrete variables in initial equation (since they are identical to y), and then the next logical step would be that initial equations are back to continuous-time equations (which automatically forbids using pre(y)).

The variants are thus (parts in parenthesis are implied by the rest of the spec):
1. during initialization pre(y)=y for non-discrete variables, initial equations are discrete, (can use pre(y) in initial equations)
2. during initialization pre(y)=y for non-discrete variables, initial equations are discrete, can NOT use pre(y) in initial equations
3. during initialization pre(y)=y for non-discrete variables, initial equations are not automatically discrete (and thus cannot use pre(y) in initial equations)

However, other good reasons to view initial equations as discrete-time expressions are:
* An initial equation for boolean of the type b=f(time, ...) requires that f(time, ...) is a discrete-time expression (3.8.3).
* An initial algorithm for a discrete-time Real such as t_start:=time; requires that time is a discrete-time expression, cp. Modelica.Blocks.Sources.SawTooth. [I assume assignment restrictions in 3.8 also apply to initial algorithms]
* It is only evaluated at one point in time.
* Relations cannot generate events during initialization; so relations in initial equations are similar to discrete-time expressions.

To me that rules out alternative (3) leaving (1) and (2).

The reason to disallow pre(y) in initial equations is that it is the same as y, and giving an equation can cause problems (especially since it was required a bit). What are the reasons for allowing pre(y) in initial equations even if pre(y)=y?
* Can describe the initialization as having initial equation pre(y)=y (instead of the during initialization pre(y)=y for non-discrete variables).
* A 'when initial()' equation is treated the same as an initial equation; this creates a simpler mental model and allows a user to separate out the initialization by copying the body of the when-clause to an initial equation - without any additional changes.

The case 'initial equation pre(y)= <constant expression>' is easy to detect for a tool, and with the rule that we have initial equation pre(y)=y I doubt that many will use it.

Thus the simplest solution seems to be (1) above - and we could just comment that using pre(y) will for a non-discrete variable gives the same value as y.

**Comment by modelica-trac-importer** [comment]
**Comment by lennart&#x2e;ochel on 10 Jul 2013 10:06 UTC**
I agree (option 1). Maybe it should be handled similar to the auto-fix mechanism that is already present for states and discrete variables. If there is a missing equation to determine v or pre(v), an equation 'pre(v)=v' should be introduced implicit (with some kind of warning). 

Replying to [comment:22 hansolsson]:
> \* A 'when initial()' equation is treated the same as an initial equation; this creates a simpler mental model and allows a user to separate out the initialization by copying the body of the when-clause to an initial equation - without any additional changes.

I do not agree. A 'when initial()' equation is very different to an initial equation. It is an equation of the simulation equation system. This can be easily verified in Dymola (or any other Modelica tool):
```mo
model Unnamed
  Real x;
equation 
  when initial() then
    x = time;
  end when;
end Unnamed;
```

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 10 Jul 2013 10:35 UTC**
Replying to [comment:23 lennart.ochel@…]:
> I agree (option 1). Maybe it should be handled similar to the auto-fix mechanism that is already present for states and discrete variables. If there is a missing equation to determine v or pre(v), an equation 'pre(v)=v' should be introduced implicit (with some kind of warning). 
> 
> Replying to [comment:22 hansolsson]:
> > \* A 'when initial()' equation is treated the same as an initial equation; this creates a simpler mental model and allows a user to separate out the initialization by copying the body of the when-clause to an initial equation - without any additional changes.
> 
> I do not agree. A 'when initial()' equation is very different to an initial equation. It is an equation of the simulation equation system. This can be easily verified in Dymola (or any other Modelica tool):
> ```mo
> model Unnamed
>   Real x;
> equation 
>   when initial() then
>     x = time;
>   end when;
> end Unnamed;
> ```

I meant the 'when initial()' part of 'when {initial(), ...}' is similar to an initial equation.

However, I realized it is not completely accurate: if 'x' is a discrete variable depending on 'pre(x)' then an initial equation and 'when initial()' will give different results.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 10 Jul 2013 12:07 UTC**
Option 1 is also fine for me.

I just would like to document one issue so that everyone is aware of it. Consider the following model (this is a simplified version of Modelica.Blocks.Logical.TriggeredTrapezoid):
```
  Real y;
  Real Tnext;
  parameter Real A=1;
  BooleanInput
equation 
  y = if time < Tnext then A - (Tnext-time) else 0;
  when {initial(), u} then
     Tnext = if initial() then time else time + A-pre(y);
  end when;
```
According to option (1), pre(y) = y during initialization and the equations during initialization are:
```
  y = if time < Tnext then A - (Tnext-time) else 0;
  Tnext = if initial() then time else time + A-pre(y);
  y = pre(y);
```
So, structurally this is a nonlinear system of equations in the unknowns, y, Tnext, pre(y).

This is probably no issue, since this system can be easily solved with one iteration. Furthermore, it is possible to rewrite the model to remove this structural dependency:
```
  Real y;
  Real Tnext(start=0, fixed=true);
  parameter Real A=1;
  BooleanInput
equation 
  y = if time < Tnext then A - (Tnext-time) else 0;
  when u then
     Tnext = time + A-pre(y);
  end when;
```
Here, during initialization only the equations
```
  Tnext = 0;
  y = if time < Tnext then A - (Tnext-time) else 0;
  pre(y) = y;
```
are present and this is structurally no equation system.

So, to summarize, minor issues can occur, but this can also be avoid by rewriting the model. So, if this rule is introduced, probably the TriggeredTrapezoidal block should be slightly changed.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 10 Jul 2013 18:35 UTC**
Discussion and decision at the web/phone meeting:

**Decision:**
Add the following to the specification (Martin will do this):

For any non-discrete variable v, conceptually the equation

```
initial equation
  pre(v) = v;
```
is present. [_If pre(v) is not present in the flattened model, a tool may not introduce this
equation, or if it was introduced it can eliminate it (to avoid the introduction of many
dummy variables pre(v))._]

**Reasoning:**
A continuous (that is non-discrete) variable should have only one value at initialization and having two (v=v.start and pre(v)) would be confusing for the user.

Comparison of discrete and non-discrete variable initialization (these are mostly the same with this definition):

Identical definition for start value and when-clause without initial()

```
   Real x(start=2, fixed=true);
   discrete Real xd(start=2, fixed=true);
equation
   der(x) = a*x;
   when sample(0,1) then
      xd = a*pre(xd) + pre(x)
   end when;

// identical initialization equations:
   x = 2
   pre(x) = 2   // since pre(x)=x and x=x.start
   pre(xd) = 2  // since pre(xd)=xd.start
   xd = 2       // since when-clause semantics: xd = pre(xd)
```
Identical definition for initial equation and when-clause without initial()

```
   Real x;
   discrete Real xd;
initial equation
   x = 2;
   xd = 2;
equation
   der(x) = a*x;
   when sample(0,1) then
      xd = a*pre(xd) + pre(x)
   end when;

// identical initialization equations:
   x = 2;
   pre(x) = 2;   // since pre(x) = x
   xd = 2;
   pre(x) = 2;   // since when-clause semantics: xd = pre(xd)
```
Different definitions if when-clause with initial()

```
   Real x;
   discrete Real xd;
initial equation
   x = 2;
   xd = 2;
equation
   der(x) = a*x;
   when {initial(), sample(0,1)} then
      xd = pre(xd) + 4*pre(x)
   end when;

// different initialization equations:
   x = 2;
   pre(x) = 2;   // since pre(x) = x
   xd = 2;
   pre(x) = -6;  // xd = pre(xd) + 4*pre(x)
                 // xd = 2
                 // -> pre(xd) = 2 - 4*pre(x) = 2-8 = -6
```
**Adaptation of MSL:**
In a few models, the initial equation "pre(x) = ..." must be removed, such as for Modelica.Blocks.Logical.TriggeredTrapezoid and for Modelica.Blocks.Math.Mean.

Martin will change MSL. It would be good if one tool vendor could introduce this rule in his engine and then check MSL, before MSL 3.2.1 is released.

**Comment by modelica-trac-importer** [comment]
**Comment by lennart&#x2e;ochel on 10 Jul 2013 20:09 UTC**
There is a typo in the last example:

Replying to [comment:26 otter]:
> ```
>    pre(x) = -6;  // xd = pre(xd) + 4\*pre(x)
>                  // xd = 2
>                  // -> pre(xd) = 2 - 4\*pre(x) = 2-8 = -6
> ```

```
   pre(xd) = -6;  // xd = pre(xd) + 4*pre(x)
                  // xd = 2
                  // -> pre(xd) = xd - 4*pre(x) = 2-8 = -6
```

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 11 Jul 2013 08:27 UTC**
Fixed in Modelica Specification in r6800:

Added in section 6.8 the following normative text:

(2) For all non-discrete (that is continuous-time) Real variables vc, the equation pre(vc) = vc is added to the initialization equations. [If pre(vc) is not present in the flattened model, a tool may not introduce this equation, or if it was introduced it can eliminate it (to avoid the introduction of many dummy variables pre(vc)).]

Added in this section the following comment:

[_Non-discrete (that is continuous-time) Real variables vc have exactly one initialization value since the rules above assure that during initialization vc = pre(vc) = vc.startExpression (if fixed= true)._

_Before the start of the integration, it must be guaranteed that for all variables v, v = pre(v). If this is not the case for some variables vi, “pre(vi) := vi” must be set and an event iteration at the initial time must follow, so the model is re-evaluated, until this condition is fulfilled._]

**Comment by modelica-trac-importer** [comment]
**Comment by pharman on 11 Jul 2013 10:11 UTC**
Replying to [comment:28 otter]:
> [If pre(vc) is not present in the flattened model, a tool may not introduce this equation, or if it was introduced it can eliminate it (to avoid the introduction of many dummy variables pre(vc)).]

Do you mean that if pre(vc) is not present the tool may choose not to introduce the equation, not that it "may not" (meaning it is not allowed to)?

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 12 Jul 2013 06:55 UTC**
Replying to [comment:26 otter]:

> **Adaptation of MSL:**
> In a few models, the initial equation "pre(x) = ..." must be removed, such as for Modelica.Blocks.Logical.TriggeredTrapezoid and for Modelica.Blocks.Math.Mean.

Opened new ticket #1207 to handle this case.

Replying to [comment:29 pharman]:
> Replying to [comment:28 otter]:
> > [If pre(vc) is not present in the flattened model, a tool may not introduce this equation, or if it was introduced it can eliminate it (to avoid the introduction of many dummy variables pre(vc)).]
> 
> Do you mean that if pre(vc) is not present the tool may choose not to introduce the equation, not that it "may not" (meaning it is not allowed to)?

Fixed the sentence in r6808.
